### ISSUE TITLE
[AppCheck] Firestore stops working after leaving the website idle overnight - AppCheck HTTP error 403 (appCheck/fetch-status-error) + infinite loop / high CPU usage

### ISSUE THREAD

**Issue by anisabboud** [firebase/firebase-js-sdk#6373]
### AppCheck problem

AppCheck works fine and enforced.
However, after leaving the webapp idle for a while (e.g., if I leave the tab open and go to sleep and put the laptop to sleep, then come to work the next day), Firestore stops working (as if it gets disconnected and doesn't reconnect or refresh AppCheck token).

Console error observed in production: `@firebase/app-check: FirebaseError: AppCheck: Fetch server returned an HTTP error status. HTTP status: 403. (appCheck/fetch-status-error).`

Console error observed on localhost: `zone.js:1061 Unhandled Promise rejection: cancelled ; Zone: <root> ; Task: Promise.then ; Value: cancelled undefined`

### Environment

  * Angular version: 14.0.2 (latest)
  * Firebase SDK version: 9.8.3 (latest)
  * AngularFire version: 7.4.1 (latest)
  * Firebase Product: AppCheck

**Comment by dconeybe** [comment]
@anisabboud Would it be possible for you to provide a minimal app to reproduce this issue, that we could use for debugging?

**Comment by anisabboud** [comment]
Hi @dconeybe, I don't currently have time to spin up a test app, but you could reproduce this issue on your own apps using Firestore listeners and AppCheck enforced. Just leave the tab open, put the computer to sleep, and check the console the next day. Try that a few days in a row.

The problem is that it's cumbersome to reproduce - I need to wait for the next day every time...
Today for instance, I encountered a different but related error when doing this - `appCheck/fetch-network-error`: \
`@firebase/app-check: FirebaseError: AppCheck: Fetch failed to connect to a network. Check Internet connection. Original error: Failed to fetch. (appCheck/fetch-network-error).`

The app continued to function, but the console was outputting the message above in an infinite loop (~10,000 times within 5 minutes).

Attaching a partial log showing the infinite loop:
![image](https://user-images.githubusercontent.com/5672108/175889288-f168821f-599a-46d7-8ac8-8aa2524cac58.png)

**Comment by dconeybe** [comment]
I've re-assigned this issue to my colleague @hsubox76 who has far more expertise in the realm of AppCheck. If this turns out to be a Firestore issue, feel free to assign it back to me.

**Comment by hsubox76** [comment]
I'm still not sure about the first error, but the second error doesn't seem to be related to the SDK itself, it seems like all the errors in the stack trace (in the last post) happen because the machine is offline, or at least the browser believes it is offline. If these happened after the computer woke up, then it seems like it either didn't reconnect to the internet properly, or the browser didn't recognize it.

As for the first error (the 403), it will be difficult to find the source of the problem if we can't reproduce it. All I can guess from my knowledge is that 403 errors like that generally happen because an invalid recaptcha token was sent to the app check exchange endpoint. A possible cause of an invalid recaptcha token might be that something about the sleep/wake process screws up the recaptcha process and causes it to yield an invalid recaptcha token, which is then sent to the app check endpoint, which returns the 403.

**Comment by anisabboud** [comment]
Hi @hsubox76 - thank you for the insights.
Note that the infinite loop of errors mentioned in my last message continues indefinitely (for more than 10 minutes, producing tens of thousands of errors until the console crashes), whereas the internet connection comes back a couple of seconds after waking the computer up.
I.e., the loop should stop when the internet comes back, but it doesn't, which seems like a bug.

![image](https://user-images.githubusercontent.com/5672108/176108539-1fe45d84-052a-4ed4-a6b1-88a652f6f0f6.png)

**Comment by clive-h-townsend** [comment]
Just chiming in here that we are experiencing the same error across a broad swath of our users. A refresh solves the issue presumably related to the reCAPTCHA expiration. For reference, we are on reCAPTCHA Enterprise. The issue appears to arise after an overnight stale session.

**Comment by anisabboud** [comment]
Posting console screenshots from today, of the errors mentioned in the first post:

### Production
> Console error observed in production: `@firebase/app-check: FirebaseError: AppCheck: Fetch server returned an HTTP error status. HTTP status: 403. (appCheck/fetch-status-error).`

In production, app continues to function, but there's an infinite loop of AppCheck errors in the console, causing high CPU usage in Chrome:
![image](https://user-images.githubusercontent.com/5672108/176847797-da0b526e-07e0-423c-b543-e86f2f62b1b7.png)

### localhost
> Console error observed on localhost: `zone.js:1061 Unhandled Promise rejection: cancelled ; Zone: <root> ; Task: Promise.then ; Value: cancelled undefined`

On localhost, no data is coming from Firestore (as if it's disconnected), but there is no infinite loop of errors in the console - just a few:
![image](https://user-images.githubusercontent.com/5672108/176848473-aba089b1-d7a8-4dd4-ba2e-acd2513d1915.png)

**Comment by clive-h-townsend** [comment]
@hsubox76 - Is there anything else you need to help identify and resolve this issue? Its difficult to replicate on our end given the time it takes for the token to expire.

**Comment by anisabboud** [comment]
Screenshot of the network tab (infinite loop 1000+ requests):

![image](https://user-images.githubusercontent.com/5672108/177169608-82f6ce8f-ace0-4968-a7e4-c385065a6352.png)

**Comment by clive-h-townsend** [comment]
Thank you for the update @hsubox76 . If we cannot ensure functionality of reCAPTCHA after a long sleep, do you have any suggestions on catching this error to fire a re-load on the client side?

**Comment by orbachar** [comment]
I had similar issue once, with the infinite errors, but it was fixed here

https://github.com/firebase/firebase-js-sdk/issues/5842

**Comment by clive-h-townsend** [comment]
Confirmed that we have the `isTokenAutoRefreshEnabled` is set to true on our app check configuration.

```
initializeAppCheck(app, {
    provider: new ReCaptchaEnterpriseProvider(XXXXXXXXXXX),
    isTokenAutoRefreshEnabled: true,
})
```

**Comment by hsubox76** [comment]
Someone is looking into why throttling isn't working.

It would be great to catch if ReCAPTCHA has gone wrong after a long sleep, and reload ReCAPTCHA if so. Unfortunately, due to the nature of ReCAPTCHA, it won't tell you on the client side if it's got a bad token or why (because otherwise bad actors could see that and do trial and error to see how to get around it), it can only be validated on the back end. So I'm not sure if there's a way to detect if ReCAPTCHA has gone stale.

**Comment by clive-h-townsend** [comment]
Do we have a timeline for when this might be deployed? As it stands, `appCheck` is causing more harm that good.

**Comment by hsubox76** [comment]
This is merged and should be released this week, usually Thursday.

**Comment by wliumelb** [comment]
#6471 
could this be related? @hsubox76

**Comment by hsubox76** [comment]
No, that would cause it to not work at all in the first place. Thanks for filing a separate issue, will address it there.

**Comment by anisabboud** [comment]
Thank you looking into this issue, and for the update.

I updated to the latest version of the Firebase SDK `9.9.1`, but still encountered some aspect of this issue / infinite loop:

Today I left my laptop idle for 1.5 hours. When I came back to it, the fans were running at full speed, indicating extreme CPU usage.
I looked at the network tab of the incognito development tab I had opened, and noticed the same behavior I mentioned in [this previous comment](https://github.com/firebase/firebase-js-sdk/issues/6373#issuecomment-1173854346). See GIF video below showing infinite network requests:

![appcheck-loop](https://user-images.githubusercontent.com/5672108/181321416-5a82ab8e-5ce5-4c70-88b3-752c80eb7297.gif)

FYI: In the video above, AuthService basically listens to the auth state via subscribing to `user(auth)` (AngularFire/RxFire syntax).
![image](https://user-images.githubusercontent.com/5672108/181339952-2863e8db-af19-45ba-8f75-7aa81d546468.png)
So something is causing auth state to be updated an infinite loop after leaving the app idle for a while.

**Comment by anisabboud** [comment]
I can confirm that the issue is still present in the new version 9.9.1.

Today I left a couple of tabs open and put the laptop to sleep.
When coming back to it the next day, the CPU was spiking like crazy:
- The localhost tabs had the infinite loop mentioned in the previous comment.
- The production tab had the following infinite loop in the console:

![AppCheck throttled high CPU usage infinite loop](https://user-images.githubusercontent.com/5672108/181700615-45fc2b6e-818b-4fc8-a730-c3134f718ec8.png)

As it stands this is very bad - it's natural for users to leave tabs open and put their laptop to sleep. When they come back to it their CPU will keep running at 100% and there is no way for them to know why their fans are running like crazy.
Even if they were tech-savvy to look into the Chrome task manager (which most people aren't), from my experience usually the high CPU usage is associated with the "Browser" process and not with a specific tab, making it extremely difficult to even identify which tab is actually spiking, unless you open the console/DevTools network tab of each tab...

I appreciate the work you've put into this so far, but I hope you could prioritize this issue so that we can reach a proper solution.
The Firebase team is heavily advertising AppCheck everywhere, and it is an excellent feature indeed, but it's currently not usable.

**Comment by anisabboud** [comment]
This just happened again on localhost while I went to have breakfast. CPU spike:
![image](https://user-images.githubusercontent.com/5672108/181709035-32679b28-f23a-4e8b-9c3b-b6903984da53.png)

I opened the Performance tab and ran profiling for 10 seconds. Infinite loop of calls while the app is idle:
![image](https://user-images.githubusercontent.com/5672108/181708561-f38dfcaa-fe67-4a47-8962-d726b87b77de.png)

Expanding one of the "Timer Fired" events shows:
![image](https://user-images.githubusercontent.com/5672108/181709434-be353086-3235-4f90-a35c-d3616557dbb9.png)

Clicking `index.esm2017.js:458:1` takes me to AppCheck code:
![image](https://user-images.githubusercontent.com/5672108/181709811-c165934d-96cc-4a70-b6c7-1e82317a6357.png)

I hope this helps.

**Comment by hsubox76** [comment]
Sorry, still looking into this. I think there's something wrong with the refresh timing. As a workaround you may be able to initialize app check with the `isTokenAutoRefreshEnabled` option set to false, and this may stop this high CPU use (at the expense of automatically refreshing the token, which may be worth it in this case, as a temporary workaround at least).
```
const appCheck = initializeAppCheck(app, {
  provider: new ReCaptchaV3Provider('abcdefghijklmnopqrstuvwxy-1234567890abcd'),
  isTokenAutoRefreshEnabled: false
});
```

**Comment by hsubox76** [comment]
I just finally looked at your gif above, and it seems to show repeated auth requests? Like the app is repeatedly logging in the user?  I know you said you didn't have time to provide a repro but I'm really curious what your app is doing, it seems like the app itself is making repeated requests, unless I'm looking at it wrong? The most recent network tab screenshot shows a lot of requests to Firestore through the webchannel, but no requests to App Check. If this was an App Check refresher problem, it wouldn't cause any repeated Auth or Firestore requests. So I'd really like to get a look at the app code.

I can't reproduce this problem. I've set up a test app with Firestore and App Check, set an `onSnapshot()` listener, tried using devtools to set the tab to "offline", changed the Firestore data on the backend to trigger snapshots, and just can't get it to make multiple repeated requests. I tried forcing a 403 error by providing a bad reCAPTCHA sitekey and it just makes requests using the correct exponential backoff times.

I don't know what angular's timer code does but it's possible it's having unexpected effects on the timeouts or promises. I think maybe taking a look at your app code would help.

**Comment by anisabboud** [comment]
Hi @hsubox76, thank you for looking into this issue!

Regarding the GIF from July 27, indeed you can see an infinite login loop.
Below the GIF you will find 3 lines of my relevant code.
You will notice that it's just standard AngularFire/RxFire code subscribing to `user(auth)` as recommended by Firebase Auth, and the code works fine normally except for these rare cases where an infinite loop is triggered.
So on that day something caused auth state to be updated an infinite loop after leaving the app idle for a while.
Most likely something expired in Firebase auth or app-check.

However, note that not all infinite loops mentioned in my posts cause the same behavior.
For example, the message I sent after that on July 29 shows an infinite loop of app-check-specific console warnings, rather than auth specific messages.

In the third message (July 29), I sent screenshots from the Chrome Performance tab which show that the infinite loop in that case is triggered in app-check code. Specifically, these files:
- https://github.com/firebase/firebase-js-sdk/blob/main/packages/app-check/src/indexeddb.ts#L96
- https://github.com/firebase/firebase-js-sdk/blob/main/packages/app-check/src/proactive-refresh.ts#L63

When you try to reproduce, try to leave the tab open overnight instead of simply setting "offline" in Network tab.
Also, from my experience, the bahavior is different on localhost vs production.
Recently, on localhost (with app-check debug token) sometimes there is no loop, but it's like Firestore completely disconnects (the app continues to function with no errors but doesn't load any data... all routes show empty data until I refresh the page).
But in production, the opposite behavior is often observed - the app continues to load data but there's an infinite loop with high CPU usage.

**Comment by anisabboud** [comment]
Example from today - loop in production:

Console tab:
![2022-08-28 appCheck console tab](https://user-images.githubusercontent.com/5672108/187080484-f0fc4503-4a1f-4361-b7d3-c1ec98e396d4.png)

Network tab:
![2022-08-28 appCheck network tab](https://user-images.githubusercontent.com/5672108/187080482-ad2ac64c-1c59-43f6-8c59-ff5c68c70e7f.png)

Internet is connected:
![image](https://user-images.githubusercontent.com/5672108/187080652-a1383334-8672-4df2-9811-d4030861fdca.png)

Related AppCheck code:
https://github.com/firebase/firebase-js-sdk/blob/f84cacd709456d4d8925343f8d6a737b7db71c61/packages/app-check/src/client.ts#L69

Not sure if this is related in some way to [Angular's Zone.js](https://angular.io/guide/zone) (don't know much about that).

**Comment by anisabboud** [comment]
This morning I encountered a **different** infinite loop in production:

This time _no requests_ in the Network tab at all, but infinite loop of warnings in the Console tab:

![image](https://user-images.githubusercontent.com/5672108/187618791-1491550b-afd8-4fe1-a962-3bae18add04e.png)

**Comment by anisabboud** [comment]
Following my last two messages, this morning I encountered a third-**different** error loop in production.
This time no login loop like Aug 28, and no infinite warnings like Aug 31, but rather infinite `appCheck/recaptcha-error` errors:

![image](https://user-images.githubusercontent.com/5672108/187855679-414b3ade-b697-4de5-adb8-5b61b722855b.png)

Following the stack trace of this particular error led me to this line of code:
https://github.com/firebase/firebase-js-sdk/blob/57b00d49fee7e900768bd0b07d38f3f8484dcfbb/packages/app-check/src/providers.ts#L73
![image](https://user-images.githubusercontent.com/5672108/187863978-eac1d9ba-6c02-49fb-8e7f-b92bc40901b1.png)
Based on my understanding of the code, this is thrown when `await getReCAPTCHAToken(this._app!)` fails. I.e., this function is failing:
https://github.com/firebase/firebase-js-sdk/blob/57b00d49fee7e900768bd0b07d38f3f8484dcfbb/packages/app-check/src/recaptcha.ts#L112

To sum up, the error loop is different each time, which is very werid and potentially indicates multiple possibly-related issues.

**Comment by skog-newglue** [comment]
I also have huge issues with this. As soon ass I leave my laptop sleeping overnight and come back the web application is not working until I refresh it. I don't however experience the high CPU usage. This is what's in my console:

![Screenshot 2022-09-02 at 08 05 17](https://user-images.githubusercontent.com/83581833/188070148-9bcb5fc6-aa68-4c61-905a-71578009cac5.png)

Please find a fix for this as it creates a horrible experience for end users using your application.

**Comment by aronsuarez** [comment]
We have exactly this problem, our emergency workaround was to turn off app check because of the bad experience with the app. 

Please find a fix

**Comment by hsubox76** [comment]
There are several different problems being mentioned in this issue. The original user is currently reporting a lot of "infinite loops" which seem like they may be caused by their app code. Since they aren't able to provide it, I can't be sure, and the logs don't provide much info other than that Firebase APIs are being called many times. I would like to look further up in the app's code to better understand what is calling them (other code in the app, or Firebase calling itself).

The other issue is that after a long sleep, browsers no longer think they are connected to the internet, and App Check does not work without an internet connection. Unfortunately from the logs provided so far for that problem specifically (not the loops), it seems like the browser itself doesn't think it's connected to the internet (ERR_INTERNET_DISCONNECTED is an error from the browser, not from Firebase). Firebase doesn't really have the ability to do anything about the browser itself not thinking it's connected to the internet.

Basically, it appears that whatever happens on this machine after a long sleep, the browser (Chrome, etc.) no longer thinks it's connected to the internet and doesn't automatically reconnect. I don't think there's anything Firebase can do about Chrome not being connected to the internet.

Given all that, what kind of fix would you like to see? Keep in mind we cannot force Chrome to reconnect to the internet, that is beyond what JS code can do. That means your app won't be able to get a valid App Check token and won't be able to access products that are gated behind App Check (whichever ones you have enabled App Check for), and if your app relies on retrieving data from App Check-gated Firestore or RTDB, it will be broken.

If you're having a problem other than one of these two ("infinite loops" or the browser itself won't reconnect by itself after a sleep) please provide the details. Particularly if the browser definitely has reconnected but App Check still is not working. It may be a different problem.

**Comment by skog-newglue** [comment]
@hsubox76 The problem is that you can still perform updates to the database, but they only seem to be updated locally in some sort of offline mode(?). So you make a bunch of changes and then you realise that your app check has made it so that it hasn't actually updated the database, despite the write requests being successful. Keep in mind here at this stage I have an internet connection. So when I'm forced to refresh the page all the changes are gone, which is a horrible user experience. 

Could a potential fix be that App Check checks your internet connection when you come back to the page after a long sleep and then performs a refresh of the token at that point if it has expired?

**Comment by hsubox76** [comment]
First, anyone who is seeing a problem reconnecting after a long sleep, can you share what OS and browser you are on?

Regarding the suggested fix above, I'm not sure I understand. If App Check checks your internet connection after you have come back after a long sleep, and it's disconnected, how can it refresh the token? I think it's trying to refresh the token while offline, which is what the problem is.

When you're saying "at this stage I have an internet connection" do you mean your computer does or your browser does? Because it seems like, at least from that screenshot, even if your computer is online, your browser (Chrome?) doesn't realize it. And if the browser doesn't know about the internet connection, there's nothing that JS code running inside the browser can do to fix that, that I know of.

The first few requests I see are Firestore trying to connect to the internet and failing, so I'm not sure why Firestore would be reporting successful writes. I might have to ask someone on the Firestore team about it. Do you have persistence enabled? https://firebase.google.com/docs/firestore/manage-data/enable-offline#configure_offline_persistence We might be able to do something in Firestore to have it report a failure instead of a success if it doesn't have a valid App Check token, even if it is writing locally, but will have to dig deeper to be sure that's what's going on first.

**Comment by anisabboud** [comment]
The internet reconnects just fine. The browser is offline only during the sleep/wake-up process until connection is regained after a couple of seconds.

I've encountered the issue that @skog-newglue mentions a few times, where the internet reconnects just fine, and the app works without errors (except for a few browser errors similar to his screenshot, which happen while the computer is being put to sleep and/or while it's waking up), but it's like Firestore gets completely disconnected. So pages that load new data would not load anything, but there are no related console errors.

**Comment by hsubox76** [comment]
I think if there are issues with Firestore specifically not reconnecting, that should probably be a separate issue, as there are too many issues combined in this thread, which I realize started out about Firestore, but is now mostly about App Check.

If there's an issue with Firestore's behavior after returning from sleep, that occurs even without App Check, I think there should be a separate issue for that. Additionally, if there's an issue with it reporting successful *writes* but failing to write to the backend, that should be another separate issue from it not getting any more live `onSnapshot()` updates.

**Comment by skog-newglue** [comment]
@hsubox76

@anisabboud Explained it perfectly. I think App Check should check if it has tried to refresh the token while being disconnected and as soon as the browser reconnects to the internet (which happens as soon as I wake up my computer) it should perform the refresh again.

I do not have Firestore persistence enabled, but I imagine App Check can use similar code to what Firestore persistence does when checking if you come back online.

I am on MacOS 10.15.7 and using Chrome Version 105.0.5195.102 (Official Build) (x86_64).

This is how I'm initiating Firestore and all related things, if it helps at all:

```
    provideFirebaseApp(() => initializeApp(environment.firebase)),
    provideAppCheck(() =>
      initializeAppCheck(getApp(), { provider: new ReCaptchaV3Provider(environment.recaptchaSiteKey), isTokenAutoRefreshEnabled: true })
    ),
    provideAuth(() => getAuth()),
    provideFirestore(() => initializeFirestore(getApp(), { ignoreUndefinedProperties: true })),
    provideStorage(() => getStorage()),
```

**Comment by hsubox76** [comment]
Looking into it. I first just want to double check that this is definitely an App Check issue and not a Firestore issue. If App Check is not enabled, you don't have any of these issues with Firestore?

**Comment by hsubox76** [comment]
To clarify how App Check and Firestore work together:

Firestore makes various requests to the backend (get, set, etc.) and if you have App Check enabled, then it first makes a call to an App Check method to get an App Check token, which it will send to the backend along with that request. If App Check fails to get that token, because it is offline, that Firestore request should fail or error (it shouldn't report a success). If we were to fix the App Check code to detect when you are back online and try to get the token again on its own, it can do that, but it can't tell Firestore to make the request again.

So I would really like to pinpoint which step the problem is happening at. Is Firestore not making a new request for an App Check token after coming back online? Is Firestore making the request but App Check won't return a valid token because it's throttled? Is Firestore making the request and App Check can't return a valid token because the long sleep has messed up reCAPTCHA?

From the screenshot it looks like Firestore hasn't been able to contact the back end at all, but as you said, the browser eventually came back online, so what requests were made after the browser came back online? Do you have a log of the network tab after the browser came back online? In an ideal situation, there should be a request to the reCAPTCHA endpoint, then the App Check endpoint, then Firestore. It would be good to see if those happened and if any of them errored.

**Comment by skog-newglue** [comment]
@hsubox76 

> If App Check is not enabled, you don't have any of these issues with Firestore?

No, the problems started to occur after I enabled it.

This is what I can see after leaving the computer sleeping overnight and waking it up the following morning.

Multiple failed requests to `POST https://securetoken.googleapis.com/v1/token?key=<my-key>` with `net::ERR_INTERNET_DISCONNECTED` status. 

Multiple failed requests to `GET https://firestore.googleapis.com/google.firestore.v1.Firestore/Listen/.....` with `net::ERR_QUIC_PROTOCOL_ERROR 200` and `net::ERR_NAME_NOT_RESOLVED`.

Multiple failed requests to `POST https://firestore.googleapis.com/google.firestore.v1.Firestore/Listen/.....` with  `net::ERR_NAME_NOT_RESOLVED`.

These happened when the computer was sleeping.


Then when it woke up it did the following:

A successful `POST https://securetoken.googleapis.com/v1/token?key=<my-key>`.

A successful `POST https://www.google.com/recaptcha/api2/reload?k=<my-key>` and after it a successful `POST https://content-firebaseappcheck.googleapis.com/v1beta/projects/.....`

I even managed to update documents in Firestore without having to refresh this time, which is.. weird.

So it seems like it already does what I proposed in the first place but it _sometimes_ fails to update the Firestore in the cloud after waking up despite successful writes.

**Comment by maccman** [comment]
I'm pretty sure this change (firebase 9.9.1) is causing infinite setTimeouts on reflect.app - we've just had to revert back to 9.9.0.

**Comment by hsubox76** [comment]
Is it possible to provide a minimal reproduction of this infinite setTimeouts issue? I'm assuming this is a new issue introduced by the "throttle fix" in 9.9.1 and not the original sleep/wake issue, and if so, can you create a new issue and provide the minimal reproduction, or at least a sample of the code that leads to these errors, or logs?

**Comment by maccman** [comment]
> Is it possible to provide a minimal reproduction of this infinite setTimeouts issue? I'm assuming this is a new issue introduced by the "throttle fix" in 9.9.1 and not the original sleep/wake issue, and if so, can you create a new issue and provide the minimal reproduction, or at least a sample of the code that leads to these errors, or logs?

I can't give you a minimal reproduction, but I can give you a version of the app with the issue if that helps?

All we did was upgrade firebase to 9.9.4. I tried each version until 9.0.0 (which fixed the issue).

**Comment by hsubox76** [comment]
Created a new issue and responded here: https://github.com/firebase/firebase-js-sdk/issues/6606

For those having trouble with Firestore sleep/wake, which specific Firestore methods are not working after wake? I know @skog-newglue mentioned writes, are these `update`, `set`, or transaction writes? Is `get` also not working? Or `onSnapshot`? Do they all report success with no errors?

**Comment by skog-newglue** [comment]
@hsubox76 For me it's `setDoc` and possibly `addDoc` as well.

**Comment by skog-newglue** [comment]
@hsubox76 Side note, I noticed today that it was firing A LOT of requests upon waking up. We're talking like 800+ requests to /token and /channel, so when I tried a write I couldn't even see the request in the network log. And what do you know, it hadn't updated the data.

So could the problem be that it's not making the request in the first place and therefore I get the impression that I've had a successful write? Just thinking out loud here. I'm probably gonna try to add a console log and try again in a few days to see if the request is actually being sent.

**Comment by hsubox76** [comment]
So if no one is able to provide a minimal repro, I understand, but it will be really hard to find and diagnose a bug without being able to reproduce it. I've created an app with only this code, nothing else (those 2 buttons at the end are in the index.html):

```
import { initializeApp } from "firebase/app";
import {
  initializeAppCheck,
  ReCaptchaV3Provider,
  onTokenChanged,
} from "firebase/app-check";
import { getFirestore, onSnapshot, doc, updateDoc } from 'firebase/firestore';

const app = initializeApp({
  /** project config */
});

async function main() {
  const appCheck = initializeAppCheck(app, {
    provider: new ReCaptchaV3Provider(
      /** recaptcha site key */
    ),
    isTokenAutoRefreshEnabled: true
  });

  onTokenChanged(appCheck, (newToken) => console.log("new token", newToken));

  const firestore = getFirestore(app);
  const docRef = doc(firestore, 'chtest/doc1');
  onSnapshot(docRef, (snap) => { console.log(snap.data())})
}

async function write() {
  const firestore = getFirestore(app);
  const docRef = doc(firestore, 'chtest/doc1');
  try {
    await updateDoc(docRef, { randomNumber: Math.round(Math.random() * 1000).toString() })
  } catch(e) {
    console.log(e);
  }
}

document.getElementById("gobutton").addEventListener("click", main);
document.getElementById("writebutton").addEventListener("click", write);
```

The buttons are in the index.html like so:
```
  <button id="gobutton">go</button>
  <button id="writebutton">write</button>
```

When you push the "gobutton" it initializes app check and starts an onSnapshot firestore listener on a certain doc. When you push the "writebutton" it writes a random number to the "randomNumber" field of that doc.
I pushed this test app to Firebase hosting and went to the url. I pushed "go" to initialize the app and "write" a few times, to write a few random numbers.
Then I put the machine to sleep overnight.
After returning in the morning there were a number of repeated Firestore errors such as `[2022-09-16T15:54:52.747Z]  @firebase/firestore: Firestore (9.10.0): Connection WebChannel transport errored: ce {type: 'c', target: Y, g: Y, defaultPrevented: false, status: 1}` and `GET https://firestore.googleapis.com/google.firestore.v1.Firestore/Listen/channel?gsessionid=flqtyYCt_guAuazxiScW1jeABUZ8MOIRW125sDNeGPY&VER=8&database=projects%2Fappcheck-testing%2Fdatabases%2F(default)&RID=rpc&SID=FAOPCVbx-faguJtrJYxdhQ&CI=0&AID=8&TYPE=xmlhttp&zx=e3rogsaw1mnv&t=1 400`
I did not see any app check errors.
When I pushed the "write" button again, however, it did a successful write, and I could see in the console that it propagated to the backend. Also, onSnapshot fired and logged the new value.
I am using Chrome on MacOS.

I think this is a pretty good foundation for a test app, if someone wants to take this code and build an app, and modify it as needed so that it reproduces the same kinds of errors you are seeing, and let me know what steps to take to reproduce it, that would be really helpful.

**Comment by hsubox76** [comment]
Was able to reproduce it by shortening the token TTL - the errors seem to happen when the token is about to expire (or has expired) and tries to auto refresh. Working on a fix in https://github.com/firebase/firebase-js-sdk/pull/6617. This seems likely to address both the rapid requests problem and the "stops working until I refresh" problem.

**Comment by hsubox76** [comment]
Sorry for the auto-close. We have a fix that should be released next week, 10/6. If you are able to test it out now, a staging version has been published (do not use in production). To get the staging version you can npm or yarn install `firebase@9.11.0-20220929164943`. If anyone is able to try this out and let me know if it works, that would be great.

**Comment by hsubox76** [comment]
Well it seems like no users were able to try the staging release but the production release is now out (9.11.0) so hopefully this fix works! I had to set up a somewhat contrived situation for the repro so I'm still not 100% sure the fix will work in users' real life situations, so let me know. I'll keep this issue open until I hear a few reports that this has been fixed in production apps for real.

**Comment by MarkDuckworth** [comment]
@anisabboud, do you have an ad blocker running which could be blocking those POST requests before and after sleep? I found others posting about the same issue and it was caused by an ad blocker.

**Comment by hsubox76** [comment]
I would suggest creating a separate issue for your Auth issues as Auth does not use App Check and while Firestore is related to Auth, it can't cause Auth sign-ins to happen. This way the Auth team can focus on that issue specifically while we can look at any possibly lingering Firestore/App Check issues here.

**Comment by anisabboud** [comment]
@MarkDuckworth I do have an ad blocker. I believe it was turned on in one tab and off in the other (different deployments of the same app), so I'm not sure it's the culprit. But will try disabling it anyway and retrying the experiment over the next couple of days. (Side note: I use windows hibernate instead of sleep.)

@hsubox76 thank you for the clarification on the difference between Auth & Firestore - will try open a separate issue if I'm able to isolate the auth issue further..

**Comment by anisabboud** [comment]
The issue I mentioned 4 days ago happens also without AdBlock.
I left two tabs open for a couple of days. One of them survived so far, but the other one ran into the issue I mentioned.

Basically that's the flow that happened on the dead tab:
1. Error POST `https://content-firebaseappcheck.googleapis.com/v1/projects/brainko1/apps/1:...:...:exchangeRecaptchaV3Token?key=...`
2. Console warning: `@firebase/app-check: AppCheck: Requests throttled due to 503 error. Attempts allowed again after 00m:01s (appCheck/throttled).`
   * This is different from the warning that I encountered 4 days ago - `Requests throttled due to 403 error. Attempts allowed again after 01d:00m:00s`, but the result is the same ↓
3. A bunch of Firestore errors on data the app is listening to (since the token refresh failed): `FirebaseError: Missing or insufficient permissions.`
4. Dysfunctional app since no data was able to load.

However, it seems that the infinite loop isn't happening lately, which is a huge relief.
Looking at the AppCheck dashboard https://console.firebase.google.com/u/0/project/_/appcheck/products,
you can notice how the number of "Unverified: invalid requests" dropped from ~10M (due to infinite loop!) to ~1K over the past few days after the update from Firebase v9.10 to v9.11.
![image](https://user-images.githubusercontent.com/5672108/195778522-8683b766-464a-49c6-b9ec-a0ad07d2d79f.png)

I'm still seeing ~1K unverified requests per day in the graph, which might be related to the issue I reported (the token expires, the token refresh fails, and then all the Firestore requests are considered invalid and fail).

**Comment by hsubox76** [comment]
Does the error you are talking about happen after you return to the tab, or while you were away?

A 503 from what I can tell indicates the server is overloaded. A 503 code causes a retry with an exponential backoff starting at about ~1s and increasing with each retry. Does this happen? You only listed one warning message. Are there any more about further attempts after that? If not, perhaps it succeeded on the second try, after 1s. Are there any successful firebaseappcheck.googleapis.com POSTs after that in the network tab?

If that's the case, Firestore should be able to resume any operations after that as normal. Which Firestore operations are you using? onSnapshot? Or one time operations? Is it failing while App Check is retrying or after it has succeeded (if it did eventually succeed)?

**Comment by anisabboud** [comment]
Q: Does the error happen after you return to the tab, or while you were away?
A: So far encountered the issue personally only 3 times since update to Firebase v9.11/9.12, while I was away from the tab.

Q: Did you see exponential backoff warnings?
A: Saw only `1d` / `1s` warnings:
- In the first instance (8 days ago - Oct 10 - running v9.11), there were a few 403 errors followed by AppCheck throttled warnings, but all of them said `"Attempts allowed again after 01d"` (1 day). I.e., they did not start from '1 sec'.
- In the second instance (4 days ago - Oct 14 - running v9.12), there was a single 503 error, with a throttled warning "1s", so your suggestion seems logical - that perhaps the following requests did not fail after 1s, but somehow the first throttling/failure caused Firestore to return `insufficient permissions` errors that broke the listeners.

Q: Which Firestore operations are you using?
A: Mostly multiple active listeners on documents & collections/queries. In a few instances one time operations.

Q: Are there any successful firebaseappcheck.googleapis.com POSTs after that in the network tab?
A: Will check once I run into the issue again. I've had two tabs open over the past few days but so far they both survived!

---

Overall the frequency of the issue was greatly reduced by the v9.11/v9.12 updates.
This is an example of a successful coming back from sleep scenario from last night:

1. I hibernate the laptop.
2. A couple of Firestore channel requests fail with `ERR_NETWORK_IO_SUSPENDED` followed by `ERR_INTERNET_DISCONNECTED`.
3. Three token requests failed `ERR_INTERNET_DISCONNECTED`
4. Internet came back and no further errors - app functions fine.
![2022-10-18 normal wake up from sleep with no issues](https://user-images.githubusercontent.com/5672108/196356492-a117c060-2d4b-4e7c-bbca-a334a425e8ce.png)
Sometimes a warning `Firestore (9.12.1): Connection WebChannel transport errored` is logged, but it doesn't cause a crash.

---

So far the issue has been correlated with the `appCheck/throttled` warning.
Perhaps Firestore should refrain from sending requests while the appCheck token is invalid/being refreshed, to avoid breaking listeners with permission errors?

---

The [AppCheck dashboard](https://console.firebase.google.com/u/0/project/_/appcheck/products) is also looking much better after your fix - currently only ~1K requests per day are labeled as "Unverified", compared to millions of requests caused by the infinite loop earlier.
![image](https://user-images.githubusercontent.com/5672108/196351602-0cb25b98-06d9-409b-914b-2f48763c12a2.png)

**Comment by hsubox76** [comment]
Yes, 403 errors are throttled for 1 day because it is likely there's something wrong with the attestation that won't be fixed with a simple retry, such as a bad API key or an attestation failure due to failing ReCAPTCHA. You can see the code for the throttling here (this also applies to 404):

https://github.com/firebase/firebase-js-sdk/blob/4eb5adfd099da3b8bd7e78bf662deda6943a8e93/packages/app-check/src/providers.ts#L286

Other errors, such as 503 server errors, could mean the server is overloaded, so it's ok to retry quickly, beginning with 1s.

If the errors are happening while the computer is asleep, that seems reasonable, as that's when the internet connection might not be working. If it doesn't work upon returning, we can focus on and address that. It seems like the main issue in this thread is fixed, if you find the issues with listener errors after waking from sleep to be a big problem, let's start a separate issue focused on that. This issue has become very long and difficult to parse as it has many related but different issues thrown together, and it seems like the biggest problem is fixed. If there's a remaining issue (listener errors after waking from sleep), let's start a separate issue that only focuses on that. (I know the original reported issue was sort of about that but this really became the "infinite loop issue" thread at some point and that's what most of the discussion is about.)
